ouroboros-consensus/test-storage/Test/Ouroboros/Storage/ChainDB/StateMachine.hs

```diff
@@ -197,25 +201,51 @@ deriving instance SOP.HasDatatypeInfo (Cmd blk it rdr)
 
 -- | Return type for successful database operations.
 data Success blk it rdr
-  = Unit           ()
-  | Chain          (AnchoredFragment (Header blk))
-  | Ledger         (ExtLedgerState blk)
-  | MbLedger       (Maybe (ExtLedgerState blk))
-  | MbBlock        (Maybe blk)
-  | MbGCedBlock    (MaybeGCedBlock blk)
-  | MbHeader       (Maybe (Header blk))
-  | Point          (Point blk)
-  | BlockNo        BlockNo
-  | UnknownRange   (UnknownRange blk)
-  | Iter           it
-  | IterResult     (IteratorResult blk)
-  | IterResultGCed (IteratorResultGCed blk)
-  | BlockReader    rdr
-  | MbChainUpdate  (Maybe (ChainUpdate blk blk))
-  | MbPoint        (Maybe (Point blk))
-  | MaxSlot        MaxSlotNo
+  = Unit                ()
+  | Chain               (AnchoredFragment (Header blk))
+  | Ledger              (ExtLedgerState blk)
+  | MbLedger            (Maybe (ExtLedgerState blk))
+  | MbBlock             (Maybe blk)
+  | MbAllComponents     (Maybe (AllComponents blk))
+  | MbGCedAllComponents (MaybeGCedBlock (AllComponents blk))
+  | MbHeader            (Maybe (Header blk))
+  | Point               (Point blk)
+  | BlockNo             BlockNo
+  | UnknownRange        (UnknownRange blk)
+  | Iter                it
+  | IterResult          (IteratorResult blk (AllComponents blk))
+  | IterResultGCed      (IteratorResultGCed blk)
+  | Rdr                 rdr
+  | MbChainUpdate       (Maybe (ChainUpdate blk (AllComponents blk)))
+  | MbPoint             (Maybe (Point blk))
+  | MaxSlot             MaxSlotNo
   deriving (Functor, Foldable, Traversable)
 
+-- | Product of all 'BlockComponent's. As this is a GADT, generating random
+-- values of it (and combinations!) is not so simple. Therefore, we just
+-- always request all block components.
+allComponents :: BlockComponent (ChainDB m blk) (AllComponents blk)
+allComponents = (,,,,,,)
+       -- TODO GetBlock and GetHeader
```

Comment:
Try to get rid of the `m` in `run` by using `traverseIterator` and `traverseReader`, etc.